### PR TITLE
 ranger: remove invalid interval when building range.

### PR DESCRIPTION
**Pull request by winoros**
If the interval is invalid to create kv range, we should remove it.
PTAL @shenli @lamxTyler @zz-jason

### CODE REVIEW

File: util/ranger/ranger.go [hunk 1]
```diff
@@ -28,6 +28,24 @@ import (
...
+		return false, errors.Trace(err)
+	}
+	if low.excl {
+		l = []byte(kv.Key(l).PrefixNext())
+	}
+	r, err := codec.EncodeKey(sc, nil, high.value)
+	if !high.excl {
+		r = []byte(kv.Key(r).PrefixNext())
+	}
+	if err != nil {
```
Comment (zz-jason) [inline util/ranger/ranger.go]: check error after `codec.EncodeKey(sc, nil, high.value)`

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -28,6 +28,24 @@ import (
 	"github.com/pingcap/tidb/util/codec"
 )
 
+func validInterval(sc *stmtctx.StatementContext, low, high point) (bool, error) {
```
Comment (lamxTyler) [inline util/ranger/ranger.go]: Why not just rewrite the `rangePointLess`?

Comment (winoros) [reply util/ranger/ranger.go]: It's also used in other case. And use encoded value is not easy to satisfy that need.

**Comment by shenli** [comment]
LGTM

File: plan/physical_plan_test.go [hunk 3]
```diff
@@ -954,8 +958,8 @@ func (s *testPlanSuite) TestRefine(c *C) {
 			best: "IndexReader(Index(t.c_d_e)[(1 3,1 +inf]])->Projection",
 		},
 		{
-			sql:  "select a from t where c in (1, 2, 3) and (d > 3 and d < 4 or d > 5 and d < 6)",
```
Comment (zz-jason) [inline plan/physical_plan_test.go]: add old test back ?

**Comment by winoros** [comment]
@zz-jason added

**Pull request merged**